File: tools/memory_cotravel_test.py
#!/usr/bin/env python3
"""Unit tests for memory_cotravel.py -- the slug-scoped agent-memory co-travel that
rides along with a transcript re-home.

Covers the two axes that make this an experimentation surface:
  * the pluggable per-file MERGE strategy (additive / source_wins / newest_mtime), each
    a pure (src,dst)->action function;
  * the FAK_MEMORY_COTRAVEL gate (off / shadow / live) -- shadow decides + ledgers but
    copies NOTHING, the prove-before-trust default.

Pure stdlib; no process spawn, no network. The ledger is redirected into tmp via
FAK_MEMORY_COTRAVEL_LEDGER so a test never touches the host store.

Run:  python -m pytest tools/memory_cotravel_test.py
"""
from __future__ import annotations

import os
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent))

import memory_cotravel as mc  # noqa: E402


# --------------------------------------------------------------------------- #
# Helpers: build a src/dst account-dir pair with a slug-scoped memory store.
# --------------------------------------------------------------------------- #
SID = "deadbeef-0000-1111-2222-333344445555"
SLUG = "C--work-fak"


def _mem_dir(cfg: str, slug: str = SLUG) -> str:
    d = os.path.join(cfg, "projects", slug, "memory")
    os.makedirs(d, exist_ok=True)
    return d


def _write(path: str, text: str) -> None:
    with open(path, "w", encoding="utf-8") as f:
        f.write(text)


def _ledger_into(tmp_path) -> str:
    p = str(tmp_path / "cotravel-ledger.jsonl")
    os.environ["FAK_MEMORY_COTRAVEL_LEDGER"] = p
    return p


def _clear_env():
    for k in ("FAK_MEMORY_COTRAVEL", "FAK_MEMORY_MERGE", "FAK_MEMORY_COTRAVEL_LEDGER"):
        os.environ.pop(k, None)


# --------------------------------------------------------------------------- #
# Pure strategy table -- no I/O beyond the byte/mtime compare.
# --------------------------------------------------------------------------- #
def test_additive_copies_only_missing(tmp_path):
    src = _mem_dir(str(tmp_path / "A"))
    dst = _mem_dir(str(tmp_path / "B"))
    _write(os.path.join(src, "note.md"), "from-A")
    # missing on dst -> copy
    assert mc._additive(os.path.join(src, "note.md"), os.path.join(dst, "note.md")) == "copy"
    # present on dst (any content) -> skip, never clobber
    _write(os.path.join(dst, "note.md"), "from-B")
    assert mc._additive(os.path.join(src, "note.md"), os.path.join(dst, "note.md")) == "skip"


def test_source_wins_overwrites_on_diff(tmp_path):
    src = _mem_dir(str(tmp_path / "A"))
    dst = _mem_dir(str(tmp_path / "B"))
    _write(os.path.join(src, "note.md"), "from-A")
    _write(os.path.join(dst, "note.md"), "from-B")
    assert mc._source_wins(os.path.join(src, "note.md"), os.path.join(dst, "note.md")) == "copy"
    # identical bytes -> no needless copy
    _write(os.path.join(dst, "note.md"), "from-A")
    assert mc._source_wins(os.path.join(src, "note.md"), os.path.join(dst, "note.md")) == "skip"


def test_newest_mtime_respects_mtime(tmp_path):
    src = _mem_dir(str(tmp_path / "A"))
    dst = _mem_dir(str(tmp_path / "B"))
    s = os.path.join(src, "note.md")
    d = os.path.join(dst, "note.md")
    _write(s, "from-A")
    _write(d, "from-B")
    os.utime(s, (3_000_000, 3_000_000))  # src newer
    os.utime(d, (1_000_000, 1_000_000))
    assert mc._newest_mtime(s, d) == "copy"
    os.utime(s, (1_000_000, 1_000_000))  # src older
    os.utime(d, (3_000_000, 3_000_000))
    assert mc._newest_mtime(s, d) == "skip"


# --------------------------------------------------------------------------- #
# Gate behavior end-to-end through cotravel_memory.
# --------------------------------------------------------------------------- #
def test_live_additive_lands_and_never_clobbers(tmp_path):
    _clear_env()
    _ledger_into(tmp_path)
    A, B = str(tmp_path / "A"), str(tmp_path / "B")
    src = _mem_dir(A)
    dst = _mem_dir(B)
    _write(os.path.join(src, "fresh.md"), "carry-me")     # missing on dst -> should land
    _write(os.path.join(src, "conflict.md"), "A-version")
    _write(os.path.join(dst, "conflict.md"), "B-version")  # exists on dst -> keep B's
    _write(os.path.join(dst, "dest-only.md"), "B-private")  # must NOT be pruned

    rec = mc.cotravel_memory(A, B, SLUG, SID, gate_value="live", strategy="additive")

    assert sorted(rec["copied"]) == ["fresh.md"]
    assert open(os.path.join(dst, "fresh.md"), encoding="utf-8").read() == "carry-me"
    # conflict kept B's bytes (never clobber)
    assert open(os.path.join(dst, "conflict.md"), encoding="utf-8").read() == "B-version"
    # dest-only survives (no prune)
    assert os.path.isfile(os.path.join(dst, "dest-only.md"))
    _clear_env()


def test_shadow_decides_but_copies_nothing(tmp_path):
    _clear_env()
    ledger = _ledger_into(tmp_path)
    A, B = str(tmp_path / "A"), str(tmp_path / "B")
    src = _mem_dir(A)
    _mem_dir(B)
    _write(os.path.join(src, "fresh.md"), "carry-me")

    rec = mc.cotravel_memory(A, B, SLUG, SID, gate_value="shadow", strategy="additive")

    # decided to copy, but in shadow NOTHING is written to the dst store
    assert rec["copied"] == []
    assert rec["would_copy"] == ["fresh.md"]
    assert not os.path.isfile(os.path.join(B, "projects", SLUG, "memory", "fresh.md"))
    # the decision IS recorded in the off-repo ledger
    assert os.path.isfile(ledger)
    rows = mc.read_ledger()
    assert rows and rows[-1]["session"] == SID and rows[-1]["gate"] == "shadow"
    _clear_env()


def test_off_is_a_noop(tmp_path):
    _clear_env()
    ledger = _ledger_into(tmp_path)
    A, B = str(tmp_path / "A"), str(tmp_path / "B")
    src = _mem_dir(A)
    _mem_dir(B)
    _write(os.path.join(src, "fresh.md"), "carry-me")

    rec = mc.cotravel_memory(A, B, SLUG, SID, gate_value="off")

    assert rec["copied"] == [] and rec["plan"] == []
    assert not os.path.isfile(os.path.join(B, "projects", SLUG, "memory", "fresh.md"))
    assert not os.path.isfile(ledger)  # off writes no ledger
    _clear_env()


def test_cross_dir_dst_slug_carries_owner_memory(tmp_path):
    # The cross-directory resume case: src memory lives under the OWNER slug, but the
    # resume runs from a different cwd whose slug is dst_slug -> the owner-slug memory
    # must land under dst_slug on the target.
    _clear_env()
    _ledger_into(tmp_path)
    A, B = str(tmp_path / "A"), str(tmp_path / "B")
    src = _mem_dir(A, SLUG)
    _write(os.path.join(src, "note.md"), "owner-memory")
    other = "C--work-slack-helpers"

    rec = mc.cotravel_memory(A, B, SLUG, SID, dst_slug=other,
                             gate_value="live", strategy="additive")

    assert rec["dst_slug"] == other
    landed = os.path.join(B, "projects", other, "memory", "note.md")
    assert os.path.isfile(landed)
    assert open(landed, encoding="utf-8").read() == "owner-memory"
    _clear_env()


def test_gate_and_strategy_default_to_shadow_additive(tmp_path):
    _clear_env()
    assert mc.gate() == "shadow"
    assert mc.strategy_name() == "additive"
    os.environ["FAK_MEMORY_COTRAVEL"] = "bogus"   # invalid -> falls back to default
    os.environ["FAK_MEMORY_MERGE"] = "bogus"
    assert mc.gate() == "shadow"
    assert mc.strategy_name() == "additive"
    _clear_env()


def test_no_source_memory_is_clean(tmp_path):
    # A session with no memory dir under the owner slug must produce an empty, harmless plan.
    _clear_env()
    _ledger_into(tmp_path)
    A, B = str(tmp_path / "A"), str(tmp_path / "B")
    os.makedirs(os.path.join(A, "projects", SLUG), exist_ok=True)  # no memory/ subdir
    rec = mc.cotravel_memory(A, B, SLUG, SID, gate_value="live")
    assert rec["src_has_memory"] is False
    assert rec["plan"] == [] and rec["copied"] == []
    _clear_env()

File: internal/memorycotravel/memorycotravel_test.go
package memorycotravel

import (
	"os"
	"path/filepath"
	"sort"
	"testing"
	"time"
)

const (
	testSID  = "deadbeef-0000-1111-2222-333344445555"
	testSlug = "C--work-fak"
)

func memDir(t *testing.T, cfg string, slug ...string) string {
	t.Helper()
	s := testSlug
	if len(slug) > 0 {
		s = slug[0]
	}
	d := filepath.Join(cfg, "projects", s, "memory")
	if err := os.MkdirAll(d, 0o755); err != nil {
		t.Fatal(err)
	}
	return d
}

func write(t *testing.T, path, text string) {
	t.Helper()
	if err := os.WriteFile(path, []byte(text), 0o644); err != nil {
		t.Fatal(err)
	}
}

func ledgerInto(t *testing.T, dir string) string {
	t.Helper()
	p := filepath.Join(dir, "cotravel-ledger.jsonl")
	t.Setenv("FAK_MEMORY_COTRAVEL_LEDGER", p)
	return p
}

func TestStrategies(t *testing.T) {
	src := memDir(t, filepath.Join(t.TempDir(), "A"))
	dst := memDir(t, filepath.Join(t.TempDir(), "B"))
	s := filepath.Join(src, "note.md")
	d := filepath.Join(dst, "note.md")
	write(t, s, "from-A")
	if Additive(s, d) != "copy" {
		t.Fatalf("additive should copy missing")
	}
	write(t, d, "from-B")
	if Additive(s, d) != "skip" {
		t.Fatalf("additive should skip existing")
	}
	if SourceWins(s, d) != "copy" {
		t.Fatalf("source wins should copy differing")
	}
	write(t, d, "from-A")
	if SourceWins(s, d) != "skip" {
		t.Fatalf("source wins should skip identical")
	}
	write(t, d, "from-B")
	if err := os.Chtimes(s, ts(3_000_000), ts(3_000_000)); err != nil {
		t.Fatal(err)
	}
	if err := os.Chtimes(d, ts(1_000_000), ts(1_000_000)); err != nil {
		t.Fatal(err)
	}
	if NewestMtime(s, d) != "copy" {
		t.Fatalf("newest should copy newer source")
	}
	if err := os.Chtimes(s, ts(1_000_000), ts(1_000_000)); err != nil {
		t.Fatal(err)
	}
	if err := os.Chtimes(d, ts(3_000_000), ts(3_000_000)); err != nil {
		t.Fatal(err)
	}
	if NewestMtime(s, d) != "skip" {
		t.Fatalf("newest should skip older source")
	}
}

func TestCotravelGates(t *testing.T) {
	t.Setenv("FAK_MEMORY_COTRAVEL", "")
	t.Setenv("FAK_MEMORY_MERGE", "")
	ledgerInto(t, t.TempDir())
	a, b := filepath.Join(t.TempDir(), "A"), filepath.Join(t.TempDir(), "B")
	src := memDir(t, a)
	dst := memDir(t, b)
	write(t, filepath.Join(src, "fresh.md"), "carry-me")
	write(t, filepath.Join(src, "conflict.md"), "A-version")
	write(t, filepath.Join(dst, "conflict.md"), "B-version")
	write(t, filepath.Join(dst, "dest-only.md"), "B-private")
	rec := CotravelMemory(a, b, testSlug, testSID, Options{Gate: "live", Strategy: "additive"})
	if got := sorted(rec.Copied); len(got) != 1 || got[0] != "fresh.md" {
		t.Fatalf("copied = %v", rec.Copied)
	}
	if data, _ := os.ReadFile(filepath.Join(dst, "conflict.md")); string(data) != "B-version" {
		t.Fatalf("conflict clobbered")
	}
	if _, err := os.Stat(filepath.Join(dst, "dest-only.md")); err != nil {
		t.Fatalf("dest-only pruned")
	}

	tmp := t.TempDir()
	ledger := ledgerInto(t, tmp)
	a, b = filepath.Join(tmp, "A"), filepath.Join(tmp, "B")
	src = memDir(t, a)
	_ = memDir(t, b)
	write(t, filepath.Join(src, "fresh.md"), "carry-me")
	rec = CotravelMemory(a, b, testSlug, testSID, Options{Gate: "shadow", Strategy: "additive"})
	if len(rec.Copied) != 0 || len(rec.WouldCopy) != 1 || rec.WouldCopy[0] != "fresh.md" {
		t.Fatalf("shadow record = %+v", rec)
	}
	if _, err := os.Stat(filepath.Join(b, "projects", testSlug, "memory", "fresh.md")); !os.IsNotExist(err) {
		t.Fatalf("shadow copied file")
	}
	if _, err := os.Stat(ledger); err != nil {
		t.Fatalf("ledger not written: %v", err)
	}
	rows := ReadLedger()
	if len(rows) == 0 || rows[len(rows)-1]["session"] != testSID || rows[len(rows)-1]["gate"] != "shadow" {
		t.Fatalf("bad ledger rows: %+v", rows)
	}

	tmp = t.TempDir()
	ledger = ledgerInto(t, tmp)
	a, b = filepath.Join(tmp, "A"), filepath.Join(tmp, "B")
	src = memDir(t, a)
	_ = memDir(t, b)
	write(t, filepath.Join(src, "fresh.md"), "carry-me")
	rec = CotravelMemory(a, b, testSlug, testSID, Options{Gate: "off"})
	if len(rec.Copied) != 0 || len(rec.Plan) != 0 {
		t.Fatalf("off should no-op: %+v", rec)
	}
	if _, err := os.Stat(ledger); !os.IsNotExist(err) {
		t.Fatalf("off wrote ledger")
	}
}

func TestDstSlugAndDefaults(t *testing.T) {
	tmp := t.TempDir()
	ledgerInto(t, tmp)
	a, b := filepath.Join(tmp, "A"), filepath.Join(tmp, "B")
	src := memDir(t, a, testSlug)
	write(t, filepath.Join(src, "note.md"), "owner-memory")
	other := "C--work-slack-helpers"
	rec := CotravelMemory(a, b, testSlug, testSID, Options{DstSlug: other, Gate: "live", Strategy: "additive"})
	if rec.DstSlug != other {
		t.Fatalf("dst slug = %s", rec.DstSlug)
	}
	if data, err := os.ReadFile(filepath.Join(b, "projects", other, "memory", "note.md")); err != nil || string(data) != "owner-memory" {
		t.Fatalf("memory did not land under dst slug: %q %v", data, err)
	}
	t.Setenv("FAK_MEMORY_COTRAVEL", "bogus")
	t.Setenv("FAK_MEMORY_MERGE", "bogus")
	if Gate() != "shadow" || StrategyName() != "additive" {
		t.Fatalf("bad defaults: %s/%s", Gate(), StrategyName())
	}
	tmp = t.TempDir()
	ledgerInto(t, tmp)
	a, b = filepath.Join(tmp, "A"), filepath.Join(tmp, "B")
	if err := os.MkdirAll(filepath.Join(a, "projects", testSlug), 0o755); err != nil {
		t.Fatal(err)
	}
	rec = CotravelMemory(a, b, testSlug, testSID, Options{Gate: "live"})
	if rec.SrcHasMemory || len(rec.Plan) != 0 || len(rec.Copied) != 0 {
		t.Fatalf("no-source-memory should be clean: %+v", rec)
	}
}

func sorted(items []string) []string {
	out := append([]string(nil), items...)
	sort.Strings(out)
	return out
}

func ts(sec int64) time.Time { return time.Unix(sec, 0) }
